### DISCUSSION TITLE
Any way to create a config for remap all cmd to ctrl?

### DISCUSSION THREAD

**Discussion by AlgusDark** [Hammerspoon/hammerspoon#3246] [Q&A] [answered]
I want to remap my cmd to be ctrl in a terminal window only. Is there an easy way to do this with hammerspoon?
I haven't found a way and I'm new with the library. 

This is something I was trying but with no success... but it works if I add a key. I want it to be any combination of modifiers and keys.

```lua
local modal = hs.hotkey.modal.new()

modal:bind({"cmd"}, 'WHAT_TO_PUT_HERE????', nil, function()
  hs.eventtap.keyStroke({"ctrl"}, 'WHAT_TO_PUT_HERE????')
end)

hs.window.filter.new('terminal')
    :subscribe(hs.window.filter.windowFocused,function()
        modal:enter()
    end)
    :subscribe(hs.window.filter.windowUnfocused,function()
        modal:exit()
    end)
```

Thanks.

**Comment by MuhammedZakir** [comment]
`hs.eventtap` is the correct one for this purpose.

E.g. https://github.com/Hammerspoon/hammerspoon/discussions/3190#discussioncomment-2610092
E.g.2. https://github.com/Hammerspoon/hammerspoon/issues/1025#issuecomment-1011435207

Edit: I didn't post the any code because I haven't done this, and don't have a Mac atm to test the code. However, I can post one if you need one; can't guarantee whether it will work though.

**Reply by AlgusDark** [reply]
Thanks @MuhammedZakir , those link helped me to get a better idea on the API, but I'm still unable to came up with a working script 😅 
It would be helpful if you could share a code and I can test/tweak it. I don't know how I could get the flags and just swap the cmd to ctrl 🤔 

So far this is my code:

```lua
local log = hs.logger.new('init.lua', 'debug')

function table.removekey(table, key)
  local element = table[key]
  table[key] = nil
  return element
end

local function keyDownHandler(event)
  local flags = event:getFlags()
  local characters = event:getCharacters(true)
  local keyCode = event:getKeyCode()
  local type = event:getType()

  if flags['cmd'] then
    table.removekey(flags, 'cmd') -- remove cmd
    table.insert(flags, 'ctrl') -- add ctrl instead
    local evt = hs.eventtap.event.newEvent() -- create a new event with the same behaviour but swapped modifier
      :setType(type)
      :setKeyCode(keyCode)
      :setFlags(flags)
    
    return true, {evt} -- this doesn't work
  end
end

local keyDownEventListener = hs.eventtap.new({
		hs.eventtap.event.types.flagsChanged,
		hs.eventtap.event.types.keyDown
	}, keyDownHandler)

hs.window.filter.new('terminal')
  :subscribe(hs.window.filter.windowFocused,function()
    keyDownEventListener:start()
  end)
  :subscribe(hs.window.filter.windowUnfocused,function()
    keyDownEventListener:stop()
  end)
```

**Comment by MuhammedZakir** [comment, marked as answer]
Notes:
1. You will have to disable "secure keyboard entry" for key remapping to work.
2. Always save objects that you want to persist indefinitely in global variables or in local variables that are referenced by global variables. Otherwise, they will be garbage collected.
3. You may move this to a module to cleanup the function/variable names.

This should work:
```lua
local eventtap = hs.eventtap
local eventTypes = eventtap.event.types
local keycodes = hs.keycodes.map
local windowFilter = hs.window.filter


local function cmd2CtrlEventHandler(event)
  local flags = event:getFlags()
  local keycode = event:getKeyCode()

  if flags.cmd then
    flags.cmd = false
    flags.ctrl = true
  end

  if keycode == keycodes.cmd then
    keycode = keycodes.ctrl
  elseif keycode == keycodes.rightcmd then
    keycode = keycodes.rightctrl
  end

  event:setKeyCode(keycode)
  event:setFlags(flags)

  return false
end

local cmd2CtrlEventTapObj = eventtap.new({
  eventTypes.flagsChanged,
  eventTypes.keyDown,
  eventTypes.keyUp
}, cmd2CtrlEventHandler)

TerminalFocusFilter = windowFilter.new('Terminal'):subscribe({
  [windowFilter.windowFocused] = function() cmd2CtrlEventTapObj:start() end,
  [windowFilter.windowUnfocused] = function() cmd2CtrlEventTapObj:stop() end
})
```

**Reply by AlgusDark** [reply]
Thanks, this definitely works.
Thanks for this 🙏

**Reply by MuhammedZakir** [reply]
Thanks! Fixed those typos.

**Reply by AlgusDark** [reply]
I noticed that `windowFilter` makes the reload of my hammerspoon to randomly hang for some time. I'm wondering if there is another way. Something I did was to have the eventap globally by doing `cmd2CtrlEventTapObj:start()`. Inside of that I added a validation of `if hs.application.frontmostApplication():name() ~= "Terminal" then return false end`.  I wonder if this is best, at least I haven't had problems when reloading my configuration now.
 
I'm asking that because I would like to rebind some buttons on slack... and I'm thinking on avoiding the windowfilter... maybe `hs.application.watcher` is a better choice?

**Reply by MuhammedZakir** [reply]
> I noticed that `windowFilter` makes the reload of my hammerspoon to randomly hang for some time.

You should open an issue about this.

> Something I did was to have the eventap globally by doing `cmd2CtrlEventTapObj:start()`. Inside of that I added a validation of `if hs.application.frontmostApplication():name() ~= "Terminal" then return false end`. I wonder if this is best, at least I haven't had problems when reloading my configuration now.

In this scenario, this seems fine, especially because you don't have any problems with this. :-)

> I'm asking that because I would like to rebind some buttons on slack... and I'm thinking on avoiding the windowfilter... maybe `hs.application.watcher` is a better choice?

Hard to say without testing. I suggest you to test them and see which works better for you.

**Reply by MuhammedZakir** [reply]
Personally, I would do:
- if I want to remap many keys that may do different things depending on some condition, I would possibly do it like how you did above.
- if I want to remap keys or do some stuff for specific apps (and only has very few to no overlaps), then I would add a watcher for apps/windows, and do stuff using callbacks like how I did in this answer.
- if what I want to do has overlaps, then I use both.

This all assumes that the performance and stability of these are similiar, or have very little difference. Otherwise: run -> test -> fix / change code -> repeat.

**Reply by AlgusDark** [reply]
Perfect, thanks. I'll do some tests :)

**Comment by zhou13** [comment]
I created the [hammerspoon-windows-mode](https://github.com/zhou13/hammerspoon-windows-mode/tree/master) using the method provided by @MuhammedZakir in case anyone needs a starting scripts to mimic keybinding in windows.
